pkg/mads/v1/reconcile/snapshot_generator.go

```diff
@@ -4,56 +4,131 @@ import (
+
+func NewSnapshotGenerator(resourceManager core_manager.ReadOnlyResourceManager, resourceGenerator generator.ResourceGenerator, meshCache *mesh.Cache) util_xds_v3.SnapshotGenerator {
 	return &snapshotGenerator{
 		resourceManager:   resourceManager,
 		resourceGenerator: resourceGenerator,
+		meshCache:         meshCache,
 	}
 }
 
 type snapshotGenerator struct {
 	resourceManager   core_manager.ReadOnlyResourceManager
 	resourceGenerator generator.ResourceGenerator
+	meshCache         *mesh.Cache
 }
 
-func (s *snapshotGenerator) GenerateSnapshot(ctx context.Context, _ *envoy_core.Node) (util_xds_v3.Snapshot, error) {
-	meshes, err := s.getMeshesWithPrometheusEnabled(ctx)
+func (s *snapshotGenerator) GenerateSnapshot(ctx context.Context, node *envoy_core.Node) (util_xds_v3.Snapshot, error) {
+	meshesWithMeshMetrics, err := s.getMeshesWithMeshMetrics(ctx, node.Id)
 	if err != nil {
 		return nil, err
 	}
 
-	dataplanes, err := s.getDataplanes(ctx, meshes)
+	meshes, err := s.getMeshesWithPrometheusEnabled(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	meshGateways, err := s.getMeshGateways(ctx, meshes)
-	if err != nil {
-		return nil, err
+	if len(meshes) > 0 && len(meshesWithMeshMetrics) > 0 {
+		log.Info("it is not supported to use both MeshMetrics policy and 'metrics' under Mesh resource. If migrating please remove the 'metrics' section and apply an equivalent MeshMetrics resource")
 	}
 
-	args := generator.Args{
-		Meshes:       meshes,
-		Dataplanes:   dataplanes,
-		MeshGateways: meshGateways,
+	var resources []*core_xds.Resource
+	if len(meshesWithMeshMetrics) == 0 {
+		dataplanes, err := s.getDataplanes(ctx, meshes)
+		if err != nil {
+			return nil, err
+		}
+
+		meshGateways, err := s.getMeshGateways(ctx, meshes)
+		if err != nil {
+			return nil, err
+		}
+
+		args := generator.Args{
+			Meshes:       meshes,
+			Dataplanes:   dataplanes,
+			MeshGateways: meshGateways,
+		}
+
+		resources, err = s.resourceGenerator.Generate(args)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		meshMetricConfToDataplanes, err := s.getMatchingDataplanes(ctx, meshesWithMeshMetrics)
+		if err != nil {
+			return nil, err
+		}
+
+		resources, err = meshmetrics_generator.Generate(meshMetricConfToDataplanes)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	resources, err := s.resourceGenerator.Generate(args)
-	if err != nil {
+	return mads_v1_cache.NewSnapshot("", core_xds.ResourceList(resources).ToIndex()), nil
+}
+
+func (s *snapshotGenerator) getMeshesWithMeshMetrics(ctx context.Context, clientId string) ([]string, error) {
+	meshMetricsList := v1alpha1.MeshMetricResourceList{}
+	if err := s.resourceManager.List(ctx, &meshMetricsList); err != nil {
 		return nil, err
 	}
 
-	return mads_v1_cache.NewSnapshot("", core_xds.ResourceList(resources).ToIndex()), nil
+	var meshToMeshMetrics []string
+	for _, meshMetric := range meshMetricsList.Items {
+		for _, backend := range *meshMetric.Spec.Default.Backends { // can backends be nil?
```

Comment:
backends can be `nil` if you have one policy for `Mesh` with backend and seperate policies for `MeshServices` that only change aggregation or sidecar config

Reply:
added check for it